x-pack/plugins/task_manager/server/kibana_discovery_service/kibana_discovery_service.ts

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ISavedObjectsRepository } from '@kbn/core/server';
+import { Logger } from '@kbn/core/server';
+import { BACKGROUND_TASK_NODE_SO_NAME } from '../saved_objects';
+import { BackgroundTaskNode } from '../saved_objects/schemas/background_task_node';
+
+interface DiscoveryServiceParams {
+  currentNode: string;
+  savedObjectsRepository: ISavedObjectsRepository;
+  logger: Logger;
+}
+
+interface DiscoveryServiceUpsertParams {
+  id: string;
+  lastSeen: string;
+}
+
+export const DISCOVERY_INTERVAL = 1000 * 10;
+export const ACTIVE_NODES_LOOK_BACK = '30s';
+
+export class KibanaDiscoveryService {
+  private currentNode: string;
+  private started = false;
+  private savedObjectsRepository: ISavedObjectsRepository;
+  private logger: Logger;
+
+  constructor({ currentNode, savedObjectsRepository, logger }: DiscoveryServiceParams) {
+    this.savedObjectsRepository = savedObjectsRepository;
+    this.logger = logger;
+    this.currentNode = currentNode;
+  }
+
+  private async upsertCurrentNode({ id, lastSeen }: DiscoveryServiceUpsertParams) {
+    await this.savedObjectsRepository.update<BackgroundTaskNode>(
+      BACKGROUND_TASK_NODE_SO_NAME,
+      id,
+      {
+        id,
+        last_seen: lastSeen,
+      },
+      { upsert: { id, last_seen: lastSeen }, refresh: false }
+    );
+  }
+
+  private async scheduleUpsertCurrentNode() {
+    const lastSeenDate = new Date();
+    const lastSeen = lastSeenDate.toISOString();
+    try {
+      await this.upsertCurrentNode({ id: this.currentNode, lastSeen });
+      if (!this.started) {
+        this.logger.info('Kibana Discovery Service has been started');
+        this.started = true;
+      }
+    } catch (e) {
+      if (!this.started) {
+        this.logger.error(
+          `Kibana Discovery Service couldn't be started and will be retried in 10s, error:${e.message}`
+        );
+      } else {
+        this.logger.error(
+          `Background Task Node couldn't be updated. id: ${this.currentNode}, last_seen: ${lastSeen}, error:${e.message}`
+        );
+      }
+    } finally {
+      setTimeout(
+        async () => await this.scheduleUpsertCurrentNode(),
+        DISCOVERY_INTERVAL - (Date.now() - lastSeenDate.getTime())
+      );
+    }
+  }
+
+  public isStarted() {
+    return this.started;
+  }
+
+  public async start() {
+    if (!this.started) {
+      await this.scheduleUpsertCurrentNode();
+    } else {
+      this.logger.warn('Kibana Discovery Service has already been started');
+    }
+  }
+
+  public async getActiveKibanaNodes() {
+    const { saved_objects: activeNodes } =
+      await this.savedObjectsRepository.find<BackgroundTaskNode>({
+        type: BACKGROUND_TASK_NODE_SO_NAME,
+        perPage: 10000,
```

Comment:
even if it should be impossible to get a high number of hits, 10k seems unnecessarily high. Maybe we can do `100` and sort in the query by the `last_seen` value so that even if it was flooded we'd still get the 100 most recent nodes. Alternatively if we want to be sure to handle all results we could actually page through the results.